src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java

```diff
@@ -0,0 +1,82 @@
+import org.geoserver.ows.AbstractDispatcherCallback;
+import org.geoserver.ows.Request;
+import org.geoserver.platform.Operation;
+import org.geoserver.platform.Service;
+import org.geoserver.platform.ServiceException;
+import org.geoserver.wfs.WFSInfo;
+import org.geotools.util.logging.Logging;
+
+/**
+ * A callback that will filter the {@link WFSGetFeatureOutputFormat} based on {@link
+ * WFSInfo#getGetFeatureOutputTypes()}
+ */
+public class WFSOutputFormatCallback extends AbstractDispatcherCallback {
+    private static final Logger LOGGER = Logging.getLogger(WFSOutputFormatCallback.class);
+    private final GeoServer geoserver;
+    private static final String WFS = "wfs";
+    private static final String GET_FEATURE = "GetFeature";
+    private static final String GET_FEATURE_WITH_LOCK = "GetFeatureWithLock";
+    public static final String INVALID_PARAMETER_VALUE = "InvalidParameterValue";
+
+    public WFSOutputFormatCallback(GeoServer geoserver) {
+        this.geoserver = geoserver;
+    }
+
+    /**
+     * Filters out Requests whose Output Format are not explicitly enabled if Get Feature Output
+     * Type Checking is enabled
+     *
+     * @param request The request.
+     * @param operation The operation for the request.
+     * @return operation if Requested Output Format is allowed
+     */
+    @Override
+    public Operation operationDispatched(Request request, Operation operation) {
+        Service service = operation.getService();
+        if (service == null
+                || !WFS.equalsIgnoreCase(service.getId())
+                || (!request.getRequest().equalsIgnoreCase(GET_FEATURE))
+                        && (!request.getRequest().equalsIgnoreCase(GET_FEATURE_WITH_LOCK))) {
+            // not a WFS service or not getFeature or not getFeatureWithLock so we are not
+            // interested in it
+            return operation;
+        }
+        WFSInfo wfs = geoserver.getService(WFSInfo.class);
+        if (!wfs.isGetFeatureOutputTypeCheckingEnabled()) {
+            // output type checking is disabled, so we are not interested in it
+            return operation;
+        }
+        // filter the output formats
+        String outputFormat = request.getOutputFormat();
+        if (outputFormat != null
+                && wfs.getGetFeatureOutputTypes() != null
+                && !wfs.getGetFeatureOutputTypes().contains(outputFormat)) {
+            LOGGER.fine(
+                    "Output Format "
+                            + outputFormat
+                            + " is not enabled for GetFeature due to Global WFS settings");
+            throw InvalidParameterException(outputFormat);
+        }
+        return operation;
+    }
+    /**
+     * Builds a ServiceException for an unallowed output format
+     *
+     * @param outputFormat the output format requested
+     * @return the ServiceException
+     */
+    public ServiceException InvalidParameterException(String outputFormat) {
```

Comment:
This method does not seem to have value:
- Not as a public method?
- By having this method the resulting service exception is raised against the wrong method.

If you really want to document this idea make a subclass of ServiceException and what you need to do in the constructor:
```java
public class OutputFormatServiceException extends ServiceException {
    public OutputFormatServiceException(String output format){
       super( "Invalid Output Format Parameter " + outputFormat, INVALID_PARAMETER_VALUE);
    }
}
```

Reply:
Jody, this is a backport of changes that were recently merged into main.  How would you like me to proceed?  Is it ok that by making the requested changes, the 2.22.x would be out of synch with main?  Should I open a new PR against main to make these changes there first?

Reply:
Oh, I see I missed my chance to review the original.

Please proceed with the backport; have a look at the proposed changes and if you agree we can make another PR.